### PR TITLE
[SPARK-22775][SQL] move dictionary related APIs from ColumnVector to WritableColumnVector

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

These dictionary related APIs are special to `WritableColumnVector` and should not be in `ColumnVector`, which will be public soon.

## How was this patch tested?

existing tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @kiszk @ueshin

**Comment by gatorsmile** [comment]
retest this please

**Review by ueshin — COMMENTED** [review]
We can simplify `reserveDictionaryIds()` at [WritableColumnVector.java#L659](https://github.com/apache/spark/blob/103dca351952c0017b013bd5ea005a2fc0e2cb50/sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/WritableColumnVector.java#L659) as well.

Otherwise, LGTM.

**Comment by cloud-fan** [comment]
retest this pease

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/WritableColumnVector.java [hunk 1]
```diff
@@ -105,6 +107,57 @@ private void throwUnsupportedException(int requiredCapacity, Throwable cause) {
   @Override
   public boolean anyNullsSet() { return anyNullsSet; }
 
+  /**
+   * Returns the dictionary Id for rowId.
+   * This should only be called when the ColumnVector is dictionaryIds.
```
Comment (kiszk) [inline sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/WritableColumnVector.java]: `ColumnVector` -> `WritableColumnVector`

Comment (kiszk) [reply sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/WritableColumnVector.java]: `dictionaryIds has WritableColumnVector`?

**Comment by kiszk** [comment]
LGTM except a few comments for wording

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**